### PR TITLE
NEW: Create class for multi-device handling

### PR DESCRIPTION
**Pull request by carterbox**
## Purpose

Simplify multiple device handling and keep code organized.

## Approach

Move the all multi-device data handling and context switching logic out of the Operators and into a new `ThreadPool` class which inherits from `concurrent.futures.ThreadPoolExecutor` and adds functions including `broadcast`, `scatter`, and `gather`. The new `ThreadPool` handles context switching for CuPy through the `map` function because you would never use the non-default context unless you are using `map`. This completely removes the need for context specific code in the Operators.

The function which divides the scan positions and data into chunks is moved to `tike.ptycho.ptycho` because it is unrelated to the fwd, adj operations. 

I have tested these changes on MONA with 2 GPUs. Technically, the multi-device code path works without GPUs now too, so we can test the multi-device logic on our CI.

### Submitter
- [x] Write a helpfully descriptive pull request title.
- [x] Organize changes into logically grouped commits with descriptive commit messages.
- [x] Document all new functions.
- [ ] Write tests for new functions or explain why they are not needed.
- [ ] Build the documentation successfully
- [x] Use [`yapf`](https://github.com/google/yapf) to format python code.

### Reviewer
- [ ] Actually read all of the code.
- [ ] Run the new code yourself.
- [ ] Write a summary of the changes as you understand them.
- [ ] Thank the submitter.

### CODE REVIEW

**Comment by carterbox** [comment]
I need to write some tests for the new pool module, but otherwise, this is ready to review.

**Comment by carterbox** [comment]
Unsure about the behavior of scatter; not sure if the input should be already split and only scattered or if scatter should do the splitting for the user.

File: src/tike/pool.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    def gather(self, x: list, worker=0, axis=0) -> np.array:
+        """Concatenate x on a single worker along the given axis."""
+        return self.xp.concatenate(
+            [self._copy_to(part, worker) for part in x],
+            axis,
+        )
+
+    def all_gather(self, x: list) -> list:
+        """Copy a scattered x to all workers."""
```
Comment (xiaodong-yu) [inline src/tike/pool.py:52]: Here is confusing. A scattered x should be already in all workers. Do you mean "Copy a gathered x to all workers."? Then, in this case, the function name should be `all_reduce`.

Comment (carterbox) [reply src/tike/pool.py:52]: Good eye! I meant `all_gather`; all threads gather x. `all_reduce` would involve taking a sum along some dimension of the array.

However, I guess the docstring should be:
```suggestion
        """Concatenate x on all worker along the given axis."""
```
and I should add the axis parameter to `all_gather`.

File: src/tike/pool.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    #     return list(self.map(self._copy_to, x, self.workers))
+
+    def gather(self, x: list, worker=0, axis=0) -> np.array:
+        """Concatenate x on a single worker along the given axis."""
+        return self.xp.concatenate(
+            [self._copy_to(part, worker) for part in x],
+            axis,
+        )
+
+    def all_gather(self, x: list) -> list:
```
Comment (carterbox) [inline src/tike/pool.py:51]:
```suggestion
    def all_gather(self, x: list, axis=0) -> list:
```

File: src/tike/pool.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        """Copy a scattered x to all workers."""
+
+        def f(worker):
+            return self.gather(x, worker)
```
Comment (carterbox) [inline src/tike/pool.py:55]:
```suggestion
        def f(worker):
            return self.gather(x, worker, axis)
```

**Pull request merged**